src/scvi/external/decipher/_trainingplan.py

```diff
@@ -0,0 +1,143 @@
+import pyro
+import torch
+
+from scvi.module.base import (
+    PyroBaseModuleClass,
+)
+from scvi.train import LowLevelPyroTrainingPlan
+
+
+class DecipherTrainingPlan(LowLevelPyroTrainingPlan):
+    """Lightning module task to train the Decipher Pyro module.
+
+    Parameters
+    ----------
+    pyro_module
+        An instance of :class:`~scvi.module.base.PyroBaseModuleClass`. This object
+        should have callable `model` and `guide` attributes or methods.
+    loss_fn
+        A Pyro loss. Should be a subclass of :class:`~pyro.infer.ELBO`.
+        If `None`, defaults to :class:`~pyro.infer.Trace_ELBO`.
+    optim
+        A Pyro optimizer instance, e.g., :class:`~pyro.optim.Adam`. If `None`,
+        defaults to :class:`pyro.optim.Adam` optimizer with a learning rate of `1e-3`.
+    optim_kwargs
+        Keyword arguments for **default** optimiser :class:`pyro.optim.Adam`.
+    """
+
+    def __init__(
+        self,
+        pyro_module: PyroBaseModuleClass,
+        loss_fn: pyro.infer.ELBO | None = None,
+        optim: pyro.optim.PyroOptim | None = None,
+        optim_kwargs: dict | None = None,
+    ):
+        super().__init__(
+            pyro_module=pyro_module,
+            loss_fn=loss_fn,
+        )
+        optim_kwargs = optim_kwargs if isinstance(optim_kwargs, dict) else {}
+        if "lr" not in optim_kwargs.keys():
+            optim_kwargs.update({"lr": 5e-3})
+        if "weight_decay" not in optim_kwargs.keys():
+            optim_kwargs.update({"weight_decay": 1e-4})
+        self.optim = pyro.optim.ClippedAdam(optim_args=optim_kwargs) if optim is None else optim
+        # We let SVI take care of all optimization
+        self.automatic_optimization = False
+
+        self.svi = pyro.infer.SVI(
+            model=self.module.model,
+            guide=self.module.guide,
+            optim=self.optim,
+            loss=self.loss_fn,
+        )
+        self.validation_step_outputs = []
+
+        # See configure_optimizers for what this does
+        self._dummy_param = torch.nn.Parameter(torch.Tensor([0.0]))
+
+    def on_validation_model_train(self):
+        """Prepare the model for validation by switching to train mode."""
+        super().on_validation_model_train()
+        if self.current_epoch > 0:
+            # freeze the batch norm layers after the first epoch
+            # 1) the batch norm layers helps with the initialization
+            # 2) but then, they seem to imply a strong normal prior on the latent space
```

Comment:
LayerNorm also has this problem? Any literature that does something similar?

Reply:
not actually sure, this was copied over from original impl, will ask